src/group_theory/is_free_group.lean

```diff
@@ -1,175 +1,134 @@
 
-private lemma lift_right_inv_aux (F : G →* X) :
-  free_group.lift.symm (F.comp (to_free_group G).symm.to_monoid_hom) = F ∘ of :=
-by { ext, simp }
+@[simp] lemma lift'_eq_free_group_lift {A : Type u} :
+  (@lift (free_group A) _ _ H _) = free_group.lift := rfl
 
-/-- A universe-polymorphic version of `is_free_group.lift'`. -/
-@[simps symm_apply]
-def lift : (generators G → X) ≃ (G →* X) :=
-{ to_fun := λ f, (free_group.lift f).comp (to_free_group G).to_monoid_hom,
-  inv_fun := λ F, F ∘ of,
-  left_inv := λ f, free_group.lift.injective begin
-    ext x,
-    simp,
-  end,
-  right_inv := λ F, begin
-    dsimp,
-    rw ←lift_right_inv_aux,
-    simp only [equiv.apply_symm_apply],
-    ext x,
-    dsimp only [monoid_hom.comp_apply, mul_equiv.coe_to_monoid_hom],
-    rw mul_equiv.symm_apply_apply,
-  end}
-
-@[ext]
-lemma ext_hom ⦃f g : G →* X⦄ (h : ∀ a, f (of a) = g (of a)) :
-  f = g :=
-is_free_group.lift.symm.injective $ funext h
-
-@[simp] lemma lift_of (f : generators G → X) (a : generators G) : (lift f) (of a) = f a :=
+@[simp] lemma lift_of (f : generators G → H) (a : generators G) : lift f (of a) = f a :=
 congr_fun (lift.symm_apply_apply f) a
 
-@[simp] lemma lift_eq_free_group_lift {A : Type u} :
-  (@lift (free_group A) H _ _ _) = free_group.lift :=
-begin
-  -- TODO: `apply equiv.symm_bijective.injective`,
-  rw [←free_group.lift.symm_symm, ←(@lift (free_group A) H _ _ _).symm_symm],
-  congr' 1,
-  ext,
-  refl,
-end
-
-/-- A universe-polymorphic version of `unique_lift`. -/
-lemma unique_lift {X : Type w} [group X] (f : generators G → X) :
-  ∃! F : G →* X, ∀ a, F (of a) = f a :=
-begin
-  have := lift.symm.bijective.exists_unique f,
-  simp_rw function.funext_iff at this,
-  exact this,
-end
+@[simp] lemma lift_symm_of (f : G →* H) (a : generators G) : (lift.symm f) a = f (of a) :=
+rfl
+
+@[ext] lemma ext_hom ⦃f g : G →* H⦄ (h : ∀ (a : generators G), f (of a) = g (of a)) : f = g :=
+lift.symm.injective (funext h)
+
+/-- The universal property of a free group: A functions from the generators of `G` to another
+group extends in a unique way to a homomorphism from `G`.
+
+Note that since `is_free_group.lift` is expressed as a bijection, it already
+expresses the universal property.
+ -/
+lemma unique_lift (f : generators G → H) : ∃! F : G →* H, ∀ a, F (of a) = f a :=
+by simpa [function.funext_iff] using (@lift G _ _ H _).symm.bijective.exists_unique f
+
+/-- If a group satisfies the universal property of a free group, then it is a free group, where
+the universal property is expressed as in `is_free_group.lift` and its properties. -/
+def of_lift {G : Type u} [group G] (X : Type u)
+  (of : X → G)
+  (lift :         ∀ {H : Type u} [group H], by exactI (X → H) ≃ (G →* H))
+  (lift_of :      ∀ {H : Type u} [group H], by exactI ∀ (f : X → H) a, lift f (of a) = f a)
+  (lift_symm_of : ∀ {H : Type u} [group H], by exactI ∀ (f : G →* H) a, lift.symm f a = f (of a)) :
+  (is_free_group G : Type (u+1)) :=
```

Comment:
Not sure how useful it is to state this step as its own lemma, rather than inline it in `is_free_group.of_unique_lift`, as it was before.